manager.go

```diff
@@ -1185,24 +1194,26 @@ func (m *jobManager) TerminateJobForUser(user string) (string, error) {
 	if err != nil {
 		return "", err
 	}
+
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
 	klog.Infof("user %q requests job %q to be terminated", user, name)
-	if err := m.stopJob(name); err != nil {
-		klog.Errorf("unable to terminate running cluster %s: %v", name, err)
+	if job, ok := m.jobs[name]; ok {
```

Comment:
putting stopjob under the lock violates the "no long running operations under the lock".  You probably need to refactor this to:

1. get the current job status under lock, then release
2. stop the job
3. if stopping is successful, take the log and update the job status

for 1 that means moving the "get job, check status" to a method with lock and defer, then here check the value.  you also want to return an error if stopErr errors out so that the user knows stop request failed.  that means then that we'll never "drop the request" until the stop job passes, but that also may mean we need additional logic here to clean up the state (i.e. you'll need to handle "job doesn't exist" and race conditions)

Reply:
ack on the time spent in the lock. I put `stopJob` inside the lock initially because it required the actual `job` metadata to determine if it was stopping a template or step job. Now that those paths share the same logic, `stopJob` does not require the `job` at all -- so I've moved it back to its original location outside the lock..
I did, however, close one race condition (`stopJob` run before prowjob is `Pending`) and return a message back to the user as requested.